### PR TITLE
Changing features argument to labels argument re: issue 313 PR

### LINKED ISSUE #313: Features Arg --> Labels Arg
In keeping with Scikit-Learn, let's change arguments called `features` to `labels`. Need to consider backward compatibility. 

This is mostly in the feature visualizers.

### PR DESCRIPTION
**Pull request by leah-kstra**
Fixes #313  

[Description]
Initial pull request to work on this during PyCon2018 sprints. 

Changes proposed in this pull request:

- Change the arguments called `features` to `labels` in feature visualizers, aligning with scikit-learn convention.
- Update documentation.

@DistrictDataLabs/team-oz-maintainers

### CODE REVIEW

**Comment by leah-kstra** [comment]
Ready for review 

@DistrictDataLabs/team-oz-maintainers

File: yellowbrick/features/jointplot.py [hunk 1]
```diff
@@ -42,9 +42,9 @@ class JointPlotVisualizer(FeatureVisualizer):
         This is inherited from FeatureVisualizer but is defined within
         JointPlotVisualizer since there are three axes objects.
 
-    feature: string, default: None
+    label: string, default: None
         The name of the X variable
-        If a DataFrame is passed to fit and feature is None, feature
+        If a DataFrame is passed to fit and label is None, label
```
Comment (bbengfort) [inline yellowbrick/features/jointplot.py]: This is another name, where the relationship between label and feature doesn't quite match up. The argument `feature` here refers to the column of the `X` matrix, in direct contrast to the `target` which is actually a label in this case! In this case, I'd recommend changing `feature` to `column`. 

See #434 (which I just created to clarify how to change this visualizer)

File: yellowbrick/features/jointplot.py [hunk 2]
```diff
@@ -176,8 +176,8 @@ def fit(self, X, y, **kwargs):
 
             if ncols > 1:
                 raise YellowbrickValueError((
-                    "X needs to be an ndarray or DataFrame with one feature, "
-                    "please select one feature from the DataFrame"
+                    "X needs to be an ndarray or DataFrame with one label, "
```
Comment (bbengfort) [inline yellowbrick/features/jointplot.py]: Here: 

```
"X needs to be an ndarray or DataFrame with one column, "
"or specify a column to select with set_params(column=x)." 
```

File: yellowbrick/features/pcoords.py [hunk 3]
```diff
@@ -53,9 +53,9 @@ def parallel_coordinates(X, y, ax=None, features=None, classes=None,
         The axis to plot the figure on. If None is passed in the current axes
         will be used (or generated if required).
 
-    features : list, default: None
-        a list of feature names to use
-        If a DataFrame is passed to fit and features is None, feature
+    labels : list, default: None
+        a list of label names to use
```
Comment (bbengfort) [inline yellowbrick/features/pcoords.py]: A list of feature labels to use
If a DataFrame is passed to fit and labels is None, feature 
labels are selected as columns of the DataFrame.

File: yellowbrick/features/pcoords.py [hunk 4]
```diff
@@ -129,9 +129,9 @@ class ParallelCoordinates(DataVisualizer):
         The axis to plot the figure on. If None is passed in the current axes
         will be used (or generated if required).
 
-    features : list, default: None
-        a list of feature names to use
-        If a DataFrame is passed to fit and features is None, feature
+    labels : list, default: None
```
Comment (bbengfort) [inline yellowbrick/features/pcoords.py]: See above comment.

File: yellowbrick/features/pcoords.py (hunk 3 above)
Comment (bbengfort) [reply yellowbrick/features/pcoords.py]: This is actually the most sensical use of the `labels` and was the motivating factor in #313

File: yellowbrick/features/radviz.py [hunk 5]
```diff
@@ -51,8 +51,8 @@ def radviz(X, y=None, ax=None, features=None, classes=None,
     ax : matplotlib Axes, default: None
         The axes to plot the figure on.
 
-    features : list of strings, default: None
-        The names of the features or columns
+    labels : list of strings, default: None
+        A list of label names.
```
Comment (bbengfort) [inline yellowbrick/features/radviz.py]: A list of feature labels.

File: yellowbrick/features/radviz.py [hunk 6]
```diff
@@ -99,9 +99,9 @@ class RadialVisualizer(DataVisualizer):
         The axis to plot the figure on. If None is passed in the current axes
         will be used (or generated if required).
 
-    features : list, default: None
-        a list of feature names to use
-        If a DataFrame is passed to fit and features is None, feature
+    labels : list, default: None
+        A list of label names to use.
```
Comment (bbengfort) [inline yellowbrick/features/radviz.py]: See above comment

File: yellowbrick/features/rankd.py [hunk 7]
```diff
@@ -53,9 +53,9 @@ def rank1d(X, y=None, ax=None, algorithm='shapiro', features=None,
     algorithm : one of {'shapiro', }, default: 'shapiro'
         The ranking algorithm to use, default is 'Shapiro-Wilk.
 
-    features : list
-        A list of feature names to use.
-        If a DataFrame is passed to fit and features is None, feature
+    labels : list
+        A list of label names to use.
+        If a DataFrame is passed to fit and labels is None, label
```
Comment (bbengfort) [inline yellowbrick/features/rankd.py]: A list of feature labels to use (see other related comments)

File: yellowbrick/features/rankd.py [hunk 8]
```diff
@@ -104,9 +104,9 @@ def rank2d(X, y=None, ax=None, algorithm='pearson', features=None,
     algorithm : one of {pearson, covariance}
         the ranking algorithm to use, default is Pearson correlation.
 
-    features : list
-        A list of feature names to use.
-        If a DataFrame is passed to fit and features is None, feature
+    labels : list
+        A list of label names to use.
+        If a DataFrame is passed to fit and labels is None, label
```
Comment (bbengfort) [inline yellowbrick/features/rankd.py]: A list of feature labels to use (see other related comments)

File: yellowbrick/features/rankd.py [hunk 9]
```diff
@@ -153,9 +153,9 @@ class RankDBase(MultiFeatureVisualizer):
     algorithm : string
         The ranking algorithm to use; options and defaults vary by subclass
 
-    features : list
-        A list of feature names to use.
-        If a DataFrame is passed to fit and features is None, feature
+    labels : list
+        A list of label names to use.
+        If a DataFrame is passed to fit and labels is None, label
```
Comment (bbengfort) [inline yellowbrick/features/rankd.py]: A list of feature labels to use (see other related comments)

File: yellowbrick/features/rankd.py [hunk 10]
```diff
@@ -300,9 +300,9 @@ class Rank1D(RankDBase):
     algorithm : one of {'shapiro', }, default: 'shapiro'
         The ranking algorithm to use, default is 'Shapiro-Wilk.
 
-    features : list
-        A list of feature names to use.
-        If a DataFrame is passed to fit and features is None, feature
+    labels : list
```
Comment (bbengfort) [inline yellowbrick/features/rankd.py]: A list of feature labels to use (see other related comments)

File: yellowbrick/features/rankd.py [hunk 11]
```diff
@@ -406,9 +406,9 @@ class Rank2D(RankDBase):
     algorithm : one of {'pearson', 'covariance'}, default: 'pearson'
         The ranking algorithm to use, default is Pearson correlation.
 
-    features : list
-        A list of feature names to use.
-        If a DataFrame is passed to fit and features is None, feature
+    labels : list
+        A list of label names to use.
```
Comment (bbengfort) [inline yellowbrick/features/rankd.py]: A list of feature labels to use (see other related comments)

File: yellowbrick/features/rankd.py [hunk 12]
```diff
@@ -449,14 +449,14 @@ class Rank2D(RankDBase):
         'covariance': lambda X: np.cov(X.transpose()),
     }
 
-    def __init__(self, ax=None, algorithm='pearson', features=None,
+    def __init__(self, ax=None, algorithm='pearson', labels=None,
                  colormap='RdBu_r', show_feature_names=True, **kwargs):
         """
         Initialize the class with the options required to rank and
-        order features as well as visualize the result.
+        order labels as well as visualize the result.
```
Comment (bbengfort) [inline yellowbrick/features/rankd.py]: Here this is actually features that are being ranked, so does not need to be changed.

File: yellowbrick/features/scatter.py [hunk 13]
```diff
@@ -57,8 +57,8 @@ def scatterviz(X,
     ax : matplotlib axes, default: None
         The axes to plot the figure on.
 
-    features : list of strings, default: None
-        The names of two features or columns.
+    labels : list of strings, default: None
+        The names of two labels or columns.
```
Comment (bbengfort) [inline yellowbrick/features/scatter.py]: The names of two features or columns (see other related comments).

File: yellowbrick/features/scatter.py [hunk 14]
```diff
@@ -113,11 +113,11 @@ class ScatterVisualizer(DataVisualizer):
             The feature name that corresponds to a column name or index postion
             in the matrix that will be plotted against the y-axis
 
-        features : a list of two feature names to use, default: None
-            List of two features that correspond to the columns in the array.
-            The order of the two features correspond to X and Y axes on the
-            graph. More than two feature names or columns will raise an error.
-            If a DataFrame is passed to fit and features is None, feature names
+        labels : a list of two label names to use, default: None
+            List of two labels that correspond to the columns in the array.
```
Comment (bbengfort) [inline yellowbrick/features/scatter.py]: A list of feature labels to use (see other related comments)

File: yellowbrick/features/scatter.py [hunk 15]
```diff
@@ -166,18 +166,18 @@ def __init__(self,
...
 
-        # Ensure with init that features doesn't have more than two features
-        if features is not None:
-            if len(features) != 2:
+        # Ensure with init that labels doesn't have more than two labels
+        if labels is not None:
+            if len(labels) != 2:
                 raise YellowbrickValueError(
-                    'ScatterVisualizer only accepts two features.')
+                    'ScatterVisualizer only accepts two labels.')
```
Comment (bbengfort) [inline yellowbrick/features/scatter.py]: 'ScatterVisualizer only accepts two feature labels.'

File: yellowbrick/features/scatter.py [hunk 16]
```diff
@@ -205,27 +205,27 @@ def fit(self, X, y=None, **kwargs):
...
+        # handle labels that are numeric columns in ndarray matrix
+        elif self.labels_ is not None and has_ndarray_int_columns(self.labels_, X):
+            f_one, f_two = self.labels_
             X_two_cols = X[:, [int(f_one), int(f_two)]]
 
         else:
             raise YellowbrickValueError("""
-                ScatterVisualizer only accepts two features, please
-                explicitly set these two features in the init kwargs or
+                ScatterVisualizer only accepts two labels, please
```
Comment (bbengfort) [inline yellowbrick/features/scatter.py]: ScatterVisualizer only accepts two feature labels, please
explicitly set these two labels or pass an array with only two columns.

File: yellowbrick/contrib/classifier/boundaries.py [hunk 17]
```diff
@@ -202,7 +202,7 @@ def __init__(self,
 
         self.x = x
         self.y = y
-        self.features_ = features
```
Comment (bbengfort) [inline yellowbrick/contrib/classifier/boundaries.py]: I'm wondering if this should be `feature_labels_` in many cases -- I'll have a larger comment for discussion towards the end.

**Comment by leah-kstra** [comment]
Thanks so much for this in-depth review! If I'm understanding your thinking here, "labels" would be an entirely graphical term. That is, "labels" are simply the names we put on the graph, while "features" are the columns / properties / part of the data. Is that right? 
If so, would it make sense to just keep the using term "features" throughout most of the code base up until we are explicitly creating the graph and labeling it? This is the approach I went for in the changes I made to the docs. 
I'm newer to python though and new to yellowbrick, so I'm still working out how `features_` is used in the code and thus I'm not sure if that approach would work everywhere. Perhaps given the structure of yellowbrick's visualizers, it doesn't make as much sense for me to think in terms of "graph-only" as I'm used to doing with other platforms. A visualizer seems to fit the data and build the graph as it goes since the end goal is a visualization. In that case, I think feature_labels would be a good way to go in order to reflect that.

**Comment by bbengfort** [comment]
@leah-kstra thanks so much for sticking with this! 

Yes, I think that's a good description; "labels" are names we put on the figure, what humans think of in reference to the properties associated with data. While "features"  are columns/properties/ part of the data, and may or may not have associated labels. Labels are essentially equivalent to `df.columns` for a pandas data frame. 

I think your intuition about visualizers is also spot on; visualizers fit the data and build the graph as it goes. If you like `feature_labels_` then I'm also good with that!

**Comment by rebeccabilbro** [comment]
Hey there @leah-kstra! Since we've had a new version release since you opened your PR and the conversation here has gotten a bit stale, I'm going to have to close this one (since your changes are already pretty far out of sync with where the library is now). Would you be willing to open a new PR off the most recent version of `develop`? We'd still love to have you contribute!

**Comment by leah-kstra** [comment]
Hey @rebeccabilbro, thanks for the heads up! I totally understand. I have been wanting to get back to this but unfortunately life has gotten in the way by being really hectic and packed the last several weeks. I'll plan to open a new PR when I can! Thanks.

**Comment by rebeccabilbro** [comment]
Awesome! Hope you get to enjoy your summer a bit ;) but we'll be looking forward to it!

**Pull request closed without merging**